libs/x-content/src/main/java/org/elasticsearch/xcontent/DotExpandingXContentParser.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.xcontent;
+
+import java.io.IOException;
+
+/**
+ * An XContentParser that reinterprets a field name containing dots as an object
+ * structure.
+ *
+ * A fieldname named 'foo.bar.baz':... will be parsed instead as 'foo':{'bar':{'baz':...}}
+ */
+public class DotExpandingXContentParser extends FilterXContentParser {
```

Comment:
I wonder if this'd be easier to use if it wrapped the original parser entirely and detected the dots in the field names itself. That way you wouldn't need to call `switchParser`, right?

Reply:
My idea was that document parsing would always wrap its parser in this up front. And any time it hit a field with a dot in the name it'd do the conversion in some local state. And if it wasn't doing dot expansion it'd just pass off to the inner parser.

Reply:
That might work, yes, let me give it a go.

We will probably want to do some kind of parser wrapping when it comes to doing the reverse of this for objects with flattened leaves, but that's explicitly something that only applies to subsections of a document, whereas this can apply to everything.